zephyr/wrapper.c

```diff
@@ -50,18 +50,23 @@ DECLARE_TR_CTX(zephyr_tr, SOF_UUID(zephyr_uuid), LOG_LEVEL_INFO);
 #define HEAP_SYSTEM_SIZE	0
 #endif
 
-/* The Zephyr heap - TODO: split heap */
-#define HEAP_SIZE	(HEAP_SYSTEM_SIZE + HEAP_RUNTIME_SIZE + HEAP_BUFFER_SIZE)
-static uint8_t __aligned(64) heapmem[HEAP_SIZE];
+/* The Zephyr heap */
+#define HEAPMEM_SIZE	(HEAP_SYSTEM_SIZE + HEAP_RUNTIME_SIZE + HEAP_BUFFER_SIZE)
+#define HEAPMEM_SHARED_SIZE	(HEAP_RUNTIME_SHARED_SIZE + HEAP_SYSTEM_SHARED_SIZE)
+
+static uint8_t __aligned(PLATFORM_DCACHE_ALIGN)heapmem[HEAPMEM_SIZE];
+static uint8_t __aligned(PLATFORM_DCACHE_ALIGN)heapmem_shared[HEAPMEM_SHARED_SIZE];
 
 /* Use k_heap structure */
 static struct k_heap sof_heap;
+static struct k_heap sof_heap_shared;
 
 static int statics_init(const struct device *unused)
 {
 	ARG_UNUSED(unused);
 
-	sys_heap_init(&sof_heap.heap, heapmem, HEAP_SIZE);
+	sys_heap_init(&sof_heap.heap, uncache_to_cache(((void *)heapmem)), HEAPMEM_SIZE);
```

Comment:
@andyross @lgirdwood So this is the reason why this is RFC. I don't think this is multicore safe in this form with heapmem put to cached memory. But basicly, if I do the reverse, let sys_heap work on uncached memory, and map the the addresses to cached in the client side only, I get frequently kernel panics (suspect the heap metadata getting corrupted by cache invalidations done by application code).

Reply:
I bet I know the problem: the memory is being allocated uncached, converted to a cached pointer, mutated, **and then freed**.  Now the uncached heap memory is reused for another block, or metadata, or something.  But then later those lines get evicted from L1 on top of the uncached region the heap is using and cause corruption.

Can you add a `z_xtensa_cache_inv()` call over the block memory being freed (or alternatively to the layer above if you know it is a cached block and won't miss anything) and see if that fixes it?

Reply:
Maybe it would still be better to keep an API like this one permanently, where the upper layer says "I want shared/uncached memory for this block" but "please return a fast/unshared block" for another allocation using a flag or whatever.  Then the heap would know to invalidate only the blocks needed because of the API used.

Reply:
Finally note that there's a portable Zephyr cache API (which is a little clumsier) which is not wired to the xtensa arch layer yet.  Don't use that, or if you do you'll need to write the stubs to make it actually work. :)   But longer term SOF probably wants to stick with portable APIs (though even then only xtensa would ever care, as the rest of the world uses coherent caches)

Reply:
@andyross can we do the cache inv in the allocator if SOC == CAVS (or similar meaning CONFIG attribute), this way application logic need not care.

Reply:
Right.  Or application logic would just know "fast/local" vs. "slow/shared" and that some platforms have performance differences.  But it also seems like the upper layer would be able to be more intelligent about invalidation than doing it for every free() call in the allocator, which also has performance benefit.  So maybe there's a line to walk.

Reply:
ok, I can put this in the SOF application logic above the Zephyr free() since I will also be allocating some compound allocations using a Linux style devm_malloc(). Easy enough done here.

Reply:
Thanks, let me check first this fixes the problem. :)

Reply:
@andyross @lgirdwood Ok the problem is trickier. Just invalidating the allocated memory is not enough.

Zephyr heap has a 4 or 8 byte chunk-header that precedes every allocation. As we have to align the public address to cacheline, the heap chunk header will be on the previous cache line. If we put zephyr heap in uncached, it has no knowledge that user will use the allocated memory via cached aliases, so heap implementation doesn't know when to validate or not.

I can hack around this by padding all allocations (to cached heap) with a long unused tail. This ensures cache flushes done on one alloc, never corrupt the heap chunk header of the following block. Basicly I guarantee that last cacheline of all allocations is unsed and application code will never access it (unless there's a bug). 

With this (plus invalidation to the actual used buffers), I can pass stress tests. But, but, this is both super ugly (as I have to make a lot of assumptions on sys_heap internals) and is wasting a great deal of memory (first alignment of start address and then the tail).

Not sure how to proceed:
  - if heap is in uncached, cached access to public buffer addresses is bound to corrupt the chunk headers 
  - if heap is cached, code is not multicore safe on platforms with noncoherent cache (like xtensa)

I'll think about this more. The latter path is potentially easier and requires less changes to Zephyr heap code (and allows SOF to reuse more). We'd need to add cache flush&invalidate that covers the heap metadata for all heap operations to ensure in multicore case the one core doesn't corrupt heap metadata used by another one. But yeah, not really thought this through, so open to ideas.

Reply:
That's exactly right: all allocations that you intended to use as cached memory must be a precise integer number of cache lines.  It's basically the same alignment requirements as DMA.  But I don't know that this is so awful.  64 bytes isn't that large, is the code really making lots of tiny objects for a local thread to use?  If so, then having per-thread heaps might not be a terrible idea.

Reply:
And just to be clear: that chunk header (assuming a shared heap) must never, ever live in a 64 byte region that enters L1 cache on any CPU.  It's a shared data structure and has to be coherent.  But as long as the SOF allocator wrapper always follows the rule above, we'll be safe.

Reply:
IIUC, it could be later HW has 128 byte cache line size.

Reply:
@andyross  @lgirdwood  @lyakh I now uploaded a version that does the correct thing as per above. We align to cacheline and also to round-up allocation sizes to ensure. But, but, this version does still fail many cases. I enabled CONFIG_ASSERT in zephyr and I'm now hitting asserts from Zephyr heap.c pointing towards corruption of heap chunk header data. I've made multiple experiments (bumping up the cacheline size, adding explicit invalidations of the datalines that host the chunk headers, etc, etc) but so far all versions hit Zephyr asserts sooner or later when stressed. 

The earlier version that accesses heap metadata via cached addresses, is still the most reliable version I have. But that is not multicore safe.

So more work is needed before merging.

Reply:
Maybe a latent buffer overrun somewhere (even a harmless one like reading 1 byte past the end of the buffer would mess this up)?  Or maybe the region the heap is in is "dirty" from some earlier pre-heap operation and there are lines in L1 that get flushed out on top of it?

An interesting test would be to put a cache flush of (the cached address of) the full heap feature before every heap operation, and enable CONFIG_SYS_HEAP_VALIDATE (which checks the full heap on every operation -- very expensive!) to see if we can catch the corruption happening in real time.

I swear I'm almost done with my own rig to run the test suite and validate the internal heap code!

Reply:
@andyross  @lgirdwood @lyakh Marking this as resolved as in latest patch I revert back to old design. Even after multiple days of intense debug, I could not pinpoint how the heap gets corrupted. Things tried for future reference:

- guard allocations at start/end of all heap allocations -> no impact, corruption still seen 
- slimming down DSP topology, aka simpler use-case -> no impact, corruption seen on all tplg variants tried (HDA, nocodec, sdw, I2S)
- bumping cacheline alignment to 128 bytes -> no impact, also made a standalone test that confirmed 64 byte is the correct value on cAVS2.5
- disabled prefetching via "xthal_set_cache_prefetch(XTHAL_PREFETCH_DISABLE);" ->no impact
- keeping DMAs stopped (ignore start for all transfers) -> significant impact to test case execution, but with enough iterations, heap corruption seen
- impact of disabling debug -> YES, has a big impact -- issue harder to hit but is still hit -- tells timing has a role
- adding explicit calls to flush and invalidate the cache area of the heap metadata at heap alloc/free -> no impact, the typical corruption is seen on the adjacent chunk (to the one alloc'ed/freed)
- running the SOF testbench and SOF cmocka audio code tests under valgrind on X86 -> no major issues found that would reveal potential out-of-bounds accesses in the SOF audio code (that could corrupt parts of the Zephyr heap)

I had multiple moments where I thought I found a cuplrit (e.g. removing some EQ-IIR calls made a huge impact to occurence rate), but in the end, none of these turned out a real fix and rather changed timings (or something) enough to make the corruption harded to hit (or just happen in difference cases).

Given difficulty of debugging this, it would seem a better path to not mix cached and uncached accesses to the same heap object. We already know if we only access the heap memory area via on uncached addresses, or only via cached addresses, all tests pass systematically. It's only the mixed address space use that triggers the issues.

So I'll resolve this conversation now.